src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -137,16 +137,32 @@ protected ImmutableArray<CompletionProvider> GetProviders(ImmutableHashSet<strin
             }
         }
 
-        protected virtual ImmutableArray<CompletionProvider> GetProviders(ImmutableHashSet<string> roles, CompletionTrigger trigger)
+        protected virtual ImmutableArray<CompletionProvider> GetProviders(
+            ImmutableHashSet<string> roles, CompletionTrigger trigger)
         {
-            if (trigger.Kind == CompletionTriggerKind.Snippets)
+            var snippetsRule = this.GetRules().SnippetsRule;
+
+            if (snippetsRule == SnippetsRule.NeverInclude)
             {
-                return GetProviders(roles).Where(p => p.IsSnippetProvider).ToImmutableArray();
+                return GetProviders(roles).Where(p => !p.IsSnippetProvider).ToImmutableArray();
             }
-            else
+            else if (snippetsRule == SnippetsRule.AlwaysInclude)
             {
                 return GetProviders(roles);
             }
+            else if (snippetsRule == SnippetsRule.IncludeAfterTypingIdentifierQuestionTab)
+            {
+                if (trigger.Kind == CompletionTriggerKind.Snippets)
+                {
+                    return GetProviders(roles).Where(p => p.IsSnippetProvider).ToImmutableArray();
+                }
+                else
+                {
+                    return GetProviders(roles).Where(p => !p.IsSnippetProvider).ToImmutableArray();
```

Comment:
It's actually any langauge with this set of options set up.  Post my change, you can have this behavior in C# or TS if you want.